### ISSUE TITLE
CI4:  Form is being redirected from POST (which is how it is generated) to a GET

### ISSUE THREAD

**Issue by SteveIreland** [opensourcepos/opensourcepos#3721]
I'm working the with Customer form at the moment.

When the form is rendered into HTML it generates (correctly I believe) a form tag with a method of POST.

However, when the Save button is pressed the POST is being permanently redirected to a GET method.  I don't understand why, nor can I figure out what is triggering that behavior.  It seems incorrect to me.

I ask because when I click on the "Save" button I'm not getting any data returned from any of the form elements (I've tried all variations of the getXXX) and suspect that the redirect behavior is the source of this problem.

This has been a showstopper for me - so some insight would be helpful.

**Comment by odiea** [comment]
I have found that adding a New Customer the save file needs to be named getSave and with the update it needs postSave. Then you can have fun with the following error.
CRITICAL - 2023-03-19 15:19:04 --> Argument 1 passed to App\Controllers\Persons::nameize() must be of the type string, null given, called in C:\wamp64\www\testci4\app\Controllers\Customers.php on line 278
in APPPATH\Controllers\Persons.php on line 56.
 1 APPPATH\Controllers\Customers.php(278): App\Controllers\Persons->nameize(null)
 2 SYSTEMPATH\CodeIgniter.php(934): App\Controllers\Customers->getSave()
 3 SYSTEMPATH\CodeIgniter.php(499): CodeIgniter\CodeIgniter->runController(Object(App\Controllers\Customers))
 4 SYSTEMPATH\CodeIgniter.php(368): CodeIgniter\CodeIgniter->handleRequest(null, Object(Config\Cache), false)
 5 FCPATH\index.php(67): CodeIgniter\CodeIgniter->run()

**Comment by SteveIreland** [comment]
Yes, renaming it to "getSave" is the problem.  It's actually a post, so when I've renamed it to "getSave" I agree it finds the function, but my problem is that the values that I retrieve are all blanks.  In the CI3 version of the system the put save method isn't being redirected to get, so it actually retrieves the post data.

On your other issue that one is resolved by prepending the string type designation with a question mark so that both nulls and string values can be passed in.

That's a frequent change that I'm making.

**Comment by odiea** [comment]
I was trying to add a new supplier but it would only save the supplier info but no people info. Claimed everything was blank no data to pass on.

**Comment by SteveIreland** [comment]
Interesting.   I just ran a test on the same action and found that save suppliers also redirects from post to get.

![image](https://user-images.githubusercontent.com/23066623/226205243-ce6ed67f-dc5e-4fd1-bdf1-85d7128b6840.png)

That doesn't happen in CI3 and shouldn't be happening in CI4 - so there is something going on that I'm not yet aware of.

**Comment by odiea** [comment]
The  suppliers update  form shows all the info correctly but will not update . I am using a previous db from years agom.LOL

**Comment by odiea** [comment]
Below is the Supplier Model save function. I do not see where it would be updating the people tables.
```
if(parent::save_value($person_data,$supplier_id))
		{
			$builder = $this->db->table('suppliers');
			if(!$supplier_id || !$this->exists($supplier_id))
			{
				$supplier_data['person_id'] = $person_data['person_id'];
				$success = $builder->insert($supplier_data);
			}
			else
			{
				$builder->where('person_id', $supplier_id);
				$success = $builder->update($supplier_data);
			}
		}
```

**Comment by SteveIreland** [comment]
I did a detailed debug on it and and found the form values aren't being passed through to the save.  Until I can resolve why the post save is hardwired to be redirected to a get save I'm going to blame that for all for the issues.  I just now tried manually setting up the routes the way that should be and it STILL get's redirected.  I'm out of time for now, but will try again when time allows.

I was just hoping that someone might recall making a change that would trigger this.

**Comment by objecttothis** [comment]
> @objecttothis @odiea 
> 
> So, today's question is: Is anyone else's POST save being redirected to a GET save?  Or is it just me?

I think I ran into that on a Postback. In other words, when the view first loads it is autorouting to the correct function, but on reload it's looking for get.  Perhaps the method set in the html form on the view is get when it should be post?

**Comment by SteveIreland** [comment]
Okay, it's good to know that it's not just me.  That eliminates a lot of red herrings that I could chase.

I confirmed that the html is being generated as "POST", but just to be sure, I also modified the form to explicitly set the method to POST and the result was it was still redirected.

Just to be sure that it wasn't something about "save", I changed the name of the functions to "persist" and tried again.  The result was that the POST persist was redirected to GET persist.

And just to see if it was something about autorouting, I added manual routing for it.  No diff.

There is something about POST that something really hates, and I just haven't yet been able to nail the blame on anything.

**Comment by odiea** [comment]
I got it to save an item today. In Secure_controller below line 83 I had to put the notorious get.
```
// this is the basic set of methods most OSPOS Controllers will implement
	public function getIndex() { return FALSE; }
	public function getSearch() { return FALSE; }
	public function getSuggest_search() { return FALSE; }
	public function getView(int $data_item_id = -1) { return FALSE; }
	public function getSave(int $data_item_id = -1) { return FALSE; }
	public function getDelete() { return FALSE; }
```

**Comment by jekkos** [comment]
All writes should be POST, otherwise it's vulnerable to CSRF.

**Comment by SteveIreland** [comment]
Okay. After extensive debugging to confirm that Code Igniter 4 didn't have a bug.  :-) I did some Googling  to determine about .htaccess could redirect a post to a git and honestly I was shocked.  How come no one ever told me that redirects have a propensity to change a get to a post?

So I confirmed that by looking at the .htaccess that this is working for 3.3.8 and found that there was an addition for CI4 that attempted to append a forward slash to the URL (which I don't believe is necessary) since I had fixed that scenario earlier via a minor tweak to the header.php file.

So I commented out those 3 or 4 lines from the .htaccess and now the post stays a post and the save is working (with some tweaks to handle another mail_chimp issue.

Today I've resolved three issues that have been dogging me for days or in one case weeks.  I'm having a great day - I hope you are too.
